include/acacia/atspi/atspi_interface.h

```diff
@@ -0,0 +1,51 @@
+#ifndef INCLUDE_ACACIA_ATSPI_ATSPI_INTERFACE_H_
+#define INCLUDE_ACACIA_ATSPI_ATSPI_INTERFACE_H_
+
+#include <iostream>
+#include <memory>
+#include <string>
+
+#include <atspi/atspi.h>
+
+namespace acacia {
+
+/**
+ * Objects descending from this class wrap an
+ * [AtspiObject](https://docs.gtk.org/atspi2/class.Object.html) of a specific
+ * type. See descendants of this class for more information about each type.
+ */
+template <typename T>
+class AtspiInterface {
+ public:
+  explicit AtspiInterface(T* interface) : interface_(interface, Deleter()) {}
+
+  AtspiInterface() {}
+  virtual ~AtspiInterface() {}
+
+  /**
+   * Tests whether the underlying interface pointer is null.
+   */
+  bool isNull() const { return !interface_; }
+
+  /**
+   * Helper function to print commonly needed values associated with this
+   * interface.
+   */
+  virtual std::string toString() const = 0;
+
+ protected:
+  class Deleter {
+   public:
+    void operator()(T* ptr) {
+      if (ptr == nullptr)
+        return;
+      g_object_unref(ptr);
```

Comment:
Great question!

Previously we were calling `Atspi[Type]Interface()` with `AtspiNode`'s `accessible_` - i.e. the `AtspiAccessible*` owned by the `AtspiNode` - with a cast to the appropriate `Atspi[Type]`. For example:

```c++
 AtspiActionInterface AtspiNode::queryAction() const {
  AtspiAction* iface = atspi_accessible_get_action_iface(accessible_);
  if (iface) {
    g_object_unref(iface);
    return AtspiActionInterface(ATSPI_ACTION(accessible_));
  }

  return AtspiActionInterface();
}
```

EXCEPT for `AtspiHyperlinkInterface`, because, inexplicably, `atspi_accessible_get_hyperlink_iface()` returns a new `AtspiHyperlink` pointer. See the [discussion](https://github.com/Igalia/acacia/pull/171#discussion_r1546985251) on Joanie's heroic fix for the issues she was having with supporting `AtspiHyperlink`.

```c++
AtspiHyperlinkInterface AtspiNode::queryHyperlink() const {
  // Unlike the other interfaces, `atspi_accessible_get_hyperlink` gives us a
  // new hyperlink object that serves as the interface. Therefore we cannot
  // free it until we're done with it. See also `atspi_hyperlink_interface.h`.
  AtspiHyperlink* hyperlink = atspi_accessible_get_hyperlink(accessible_);
  if (hyperlink) {
    return AtspiHyperlinkInterface(hyperlink);
  }

  return AtspiHyperlinkInterface();
}
```

In the example above, `iface` is a pointer to the same object as `accessible_`, and the `atspi_accessible_get_action_iface()` function increases its ref count - hence immediately calling `g_object_unref(iface)` before calling the `AtspiActionInterface` constructor with `accessible_`.

However, that strategy means the `AtspiActionInterface` is only useful as long as `accessible_` hasn't been `free`d (AFAICT we were also never `unref`ing `accessible_` before this change, so it wouldn't have been an issue, but I'm not sure how that wasn't showing up as a memory leak...)

With this change:
- We can treat `AtspiHyperlink` the same as every other interface
- `Atspi[Type]Interface` objects' lifetimes aren't dependent on the `AtspiNode`'s lifetime.

Reply:
Thanks for the explanation! And glad you removed that interface dependency on the main node, I'm sure that would eventually have caused headaches.